### DISCUSSION TITLE
NextJS + Samlify (Route Handler Request Object Extended Issue)

### DISCUSSION THREAD

**Discussion by boncz92** [vercel/next.js#59681] [App Router]
When im attempting to migrate from nextjs page router to the app router im encountering an error. It seems that the error is being caused by the change in how the 'request' object is handled prior to getting ahold of it. The first example returns parsed results as expected while the second example returns an error:

' TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received undefined'

This is obviously a way in which samlify is handling the parsing of the request object passed to it (There isnt an option to just pass it the 'samlResponse') it expects a normally structured unaltered response object passed to it. Is there a way to retreive the raw request without the extended functionality of NextRequest? If not anyone have any ideas how I would go about resolving this? (Examples have been validated and the first example 100% validates and outputs the parsedResults)

/pages/api/auth/samlAuth.js

    import fs from 'fs';
    import * as saml from 'samlify';
    import * as saml_validator from '@authenio/samlify-validate-with-xmllint';

    const ServiceProvider = saml.ServiceProvider;
    const IdentityProvider = saml.IdentityProvider;

    const SAMLMetaDataSP = fs.readFileSync(process.env.SAMLMetaDataSP);
    const SAMLMetaDataIDP = fs.readFileSync(process.env.SAMLMetaDataIDP);

    const samlSP = ServiceProvider({
        metadata: SAMLMetaDataSP,
    });

    const samlIDP = IdentityProvider({
        metadata: SAMLMetaDataIDP,
    });

    saml.setSchemaValidator(saml_validator);

    export default async function handler(req, res) {
        const method = req.method

    let UserInformation = {};
        await samlSP.parseLoginResponse(samlIDP, 'post', req)
            .then(async parsedResults => {
                console.log(parsedResults)
            })
            .catch(err => {
                console.log(err)
                console.log("Error")
            })

        return res.json({ UserInformation })
    }

/app/api/samlAuth/route.js

    import fs from 'fs';
    import * as saml from 'samlify';
    import * as saml_validator from '@authenio/samlify-validate-with-xmllint';

    const ServiceProvider = saml.ServiceProvider;
    const IdentityProvider = saml.IdentityProvider;

    const SAMLMetaDataSP = fs.readFileSync(process.env.SAMLMetaDataSP);
    const SAMLMetaDataIDP = fs.readFileSync(process.env.SAMLMetaDataIDP);

    const samlSP = ServiceProvider({
        metadata: SAMLMetaDataSP,
    });

    const samlIDP = IdentityProvider({
        metadata: SAMLMetaDataIDP,
    });

    saml.setSchemaValidator(saml_validator);

    export async function POST(request) {
        const method = request.method
        const referer = request.headers.get('referer')

        let UserInformation = {};
        await samlSP.parseLoginResponse(samlIDP, 'post', request)
            .then(async parsedResults => {
                console.log(parsedResults)
            })
            .catch(err => {
                console.log(err)
                console.log("Error")
            })

        return Response.json({ UserInformation })
    }

    export async function GET() {
        const data = { Test: true }

        return Response.json({ data })
    }

**Comment by Pruxis** [comment]
Currently facing the same issue, I think it's with how the request is being passed to parseLoginResponse. Wondering if you ended up solving it

**Comment by Pruxis** [comment]
I managed to solve this by using the formData API.
Just pass it as json to the parseLoginResponse.

```
  const formData = await req.formData();
    const samlResponse = formData.get("SAMLResponse");
    if (!samlResponse || typeof samlResponse !== "string") {
      console.error("[saml] No SAMLResponse found in form data");
      return new NextResponse(null, { status: 400 });
    }
    const result = await serviceProvider.parseLoginResponse(identityProvider, "post", {
      body: {
        SAMLResponse: samlResponse,
      },
    });
```
